### PR TITLE
feat/npm

### PR DESCRIPTION
**Pull request by rowanc1**
See #8 

This is a draft of some of the changes. I also have recently been testing and setting up github demos and testing through actions. I can pull that together if you want.

I don't think we really need to move away from parcel? This will be used inside of other systems so a `tsc` should be good enough!?

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -7,7 +7,7 @@
 		<link rel="stylesheet" href="./node_modules/prosemirror-view/style/prosemirror.css">
 		<link rel="stylesheet" href="style/main.css">
 		<link rel="stylesheet" href="style/math.css">
-		<script src="./src/index.ts"></script>
+		<script src="./index.ts"></script>
```
Comment (rowanc1) [inline index.html:10]: I have moved the demo to the main folder. and then started to replace the index file with exports. More work necessary there.

File: package.json [hunk 2]
```diff
@@ -2,8 +2,11 @@
   "name": "prosemirror-math",
   "version": "0.0.1",
   "description": "Schema and plugins for first-class math support in ProseMirror.",
-  "main": "src/index.js",
+  "main": "dist/index.js",
+  "types": "dist/index.d.ts",
```
Comment (rowanc1) [inline package.json:6]: This is important to allow downstream packages type access.

File: package.json [hunk 3]
```diff
@@ -2,8 +2,11 @@
   "name": "prosemirror-math",
   "version": "0.0.1",
   "description": "Schema and plugins for first-class math support in ProseMirror.",
-  "main": "src/index.js",
+  "main": "dist/index.js",
```
Comment (rowanc1) [inline package.json:5]: This is what is actually published.

File: package.json [hunk 4]
```diff
@@ -2,8 +2,11 @@
...
+  "types": "dist/index.d.ts",
   "scripts": {
+    "clean": "rm -rf dist",
+    "prepublishOnly": "npm run clean; tsc; cp style/math.css dist/math.css",
```
Comment (rowanc1) [inline package.json:9]: Added something that just compiles and then copies over the css into the dist folder.

File: package.json [hunk 5]
```diff
@@ -15,7 +18,8 @@
   ],
   "author": "Benjamin R. Bray",
   "license": "MIT",
-  "dependencies": {
+  "dependencies": {},
+  "peerDependencies": {
```
Comment (rowanc1) [inline package.json:22]: Moved everything to peer and dev deps. Need both!

File: src/index.ts [hunk 6]
```diff
@@ -18,17 +18,17 @@ import { mathInputRules } from "./plugins/math-inputrules";
 import { editorSchema } from "./math-schema";
 import { MathView, ICursorPosObserver } from "./math-nodeview";
 import { mathBackspace } from "./plugins/math-backspace";
-import { mathPlugin } from "./math-plugin";
+import { mathPlugin, createMathView } from "./math-plugin";
```
Comment (rowanc1) [inline src/index.ts:21]: My schema was different than yours, so I actually exported the createMathView -- which might be better? But the css and things depended on the naming of the nodes.

File: src/index.ts [hunk 7]
```diff
@@ -44,33 +44,31 @@ function insertMath(){
 	}
 }
 
-function initEditor(){
-	// get editor element
-	let editorElt = document.getElementById("editor");
-	if(!editorElt){ throw Error("missing #editor element"); }
+// function initEditor(){
```
Comment (rowanc1) [inline src/index.ts:47]: Just commented out for now ..

File: src/math-nodeview.ts [hunk 8]
```diff
@@ -38,7 +38,7 @@ export class MathView implements NodeView, ICursorPosObserver {
 	private _getPos: (() => number);
 
 	// nodeview dom
-	dom: HTMLElement;
+	dom?: HTMLElement;
```
Comment (rowanc1) [inline src/math-nodeview.ts:41]: My TSC complained at this from the delete. There might be a better way to do this?

File: src/math-plugin.ts [hunk 9]
```diff
@@ -33,8 +33,8 @@ function createMathView(displayMode:boolean){
 
 		// set up NodeView
 		let nodeView = new MathView(
-			node, view, getPos as (() => number), 
-			{ katexOptions : { displayMode, macros: pluginState.macros } },
+			node, view, getPos as (() => number),
+			{ tagName: displayMode ? 'math-display' : 'math-inline', katexOptions : { displayMode, macros: pluginState.macros } },
```
Comment (rowanc1) [inline src/math-plugin.ts:37]: I keyed this off of the display mode, which makes it more explicit and self contained?

File: tsconfig.json [hunk 10]
```diff
@@ -8,11 +8,11 @@
     // "allowJs": true,                       /* Allow javascript files to be compiled. */
     // "checkJs": true,                       /* Report errors in .js files. */
     // "jsx": "preserve",                     /* Specify JSX code generation: 'preserve', 'react-native', or 'react'. */
-    // "declaration": true,                   /* Generates corresponding '.d.ts' file. */
+    "declaration": true,                      /* Generates corresponding '.d.ts' file. */
```
Comment (rowanc1) [inline tsconfig.json:11]: Export the .d.ts files!

File: tsconfig.json [hunk 11]
```diff
@@ -8,11 +8,11 @@
...
     // "declarationMap": true,                /* Generates a sourcemap for each corresponding '.d.ts' file. */
     // "sourceMap": true,                     /* Generates corresponding '.map' file. */
     // "outFile": "./",                       /* Concatenate and emit output to single file. */
-    // "outDir": "./",                        /* Redirect output structure to the directory. */
+    "outDir": "./dist",                        /* Redirect output structure to the directory. */
```
Comment (rowanc1) [inline tsconfig.json:15]: Put everything in dist!

File: tsconfig.json [hunk 12]
```diff
@@ -64,5 +64,6 @@
     /* Experimental Options */
     // "experimentalDecorators": true,        /* Enables experimental support for ES7 decorators. */
     // "emitDecoratorMetadata": true,         /* Enables experimental support for emitting type metadata for decorators. */
-  }
+  },
+  "exclude": ["./index.ts"]
```
Comment (rowanc1) [inline tsconfig.json:68]: It complains it there are ts files outside, so i just ignored the demo.

**Review by rowanc1 — COMMENTED** [review]
These might help -- let me know if you want me to take this over the line!

**Comment by benrbray** [comment]
Thanks for the draft PR!  It was a useful reference while preparing the npm release.  You can now `npm install benrbray@prosemirror-math`.  I also [exported](https://github.com/benrbray/prosemirror-math/blob/master/src/math-plugin.ts#L24) `createMathView`.  Let me know if there are any other changes I can make to the public api to improve usability.

**Comment by rowanc1** [comment]
Nice work!!

**Pull request closed without merging**